### PR TITLE
[NG] Datagrid numeric filter

### PR DESCRIPTION
**Pull request by amellnik**
This is a first pass at https://github.com/vmware/clarity/issues/2908.  It's not yet complete (no tests, minimal documentation, some bugs), but I was hoping to solicit some early feedback.  

image

The idea is to add a second built-in filter type that's similar in scope to the existing string filter.  Rather than a single input this second filter has two input fields which can optionally contain a lower and upper limit respectively.  This is currently called a numeric filter and uses number input fields, but a better name for it might be a range filter, since it would be easy for it to apply to non-numerical types (for example, a user could input letters to select a range alphabetically).  

There's demos for it in the datagrid section of the KS app in the "binding model properties" and the "built-in filters" sections.  

Open issues:
-  ~Rename to "range" filters and allow non-numeric inputs?~
- [x] Why is the filter icon on the wrong side?!?
- [x] Make the template look nicer
- [x] `ClrDatagridColumn` currently is closely tied to string filters ([example](https://github.com/amellnik/clarity/blob/topic/datagrid-numeric-filter/src/clr-angular/data/datagrid/datagrid-column.ts#L84))
- [x] Tests, specs
- [x] Docs
- [x] No hardcoded strings in template
- [x] Actual UI design

### CODE REVIEW

**Comment by amellnik** [comment]
Hi @gnomeontherun,

Thanks for the early feedback!  Two further questions:

* Do you think it would be worth generalizing this to other types of range filters or would it be better to stick with numeric values only?  
* Would you be open to adding a `value(): any` method to `ClrDatagridFilterInterface`?  If we expect all filters to expose some indication of their state I that would allow me to get rid of a lot of the `if` blocks to special case the two built-in filters.

File: src/dev/src/app/datagrid/binding-properties/binding-properties.html [hunk 1]
```diff
@@ -16,20 +16,27 @@ <h2>Binding model properties to columns</h2>
...
+  By default, bound columns are assumed to contain string-like contents and the user is presented with the normal string filter.
+  If you know that the contents of the column will be numeric, you can instead use the built-in numeric range filter by adding
+  <code class="clr-code">[clrDgColType]=&quot;'number'&quot;</code>.  You can see an example of this in the ID column.
 </p>
 
 <clr-datagrid>
-  <clr-dg-column>User ID</clr-dg-column>
-  <clr-dg-column [clrDgField]="'name'" [(clrFilterValue)]="nameFilter">Name</clr-dg-column>
   <clr-dg-column [clrDgField]="'creation'">Creation date</clr-dg-column>
+  <clr-dg-column [clrDgField]="'id'" [clrDgColType]="'number'">User ID</clr-dg-column>
```
Comment (youdz) [inline src/dev/src/app/datagrid/binding-properties/binding-properties.html]: I forgot these weren't cleaned up. Do you mind using the "simpler" syntax without `[...]` binding?
For this column, it would be:
```html
<clr-dg-column clrDgField="id" clrDgColType="number">User ID</clr-dg-column>
```
It gets rid of these nasty nested quotes.

File: src/clr-angular/data/datagrid/index.ts [hunk 2]
```diff
@@ -21,11 +21,14 @@ export * from './datagrid-placeholder';
 
 export * from './interfaces/state.interface';
 export * from './enums/sort-order.enum';
+export * from './enums/column-type.enum';
```
Comment (youdz) [inline src/clr-angular/data/datagrid/index.ts]: I don't see the consumer using this enum in any of your examples, can we just keep it private?

File: src/clr-angular/data/datagrid/datagrid-column.ts [hunk 3]
```diff
@@ -144,6 +153,30 @@ export class ClrDatagridColumn<T = any> extends DatagridFilterRegistrar<T, Datag
...
+  /*
+     * What type is this column?  This defaults to STRING, but can also be
+     * set to NUMBER.  Unsupported types default to STRING. Users can set it
+     * via the [clrDgColType] input by setting it to 'string' or 'number'.
+     */
+
+  private _colType: ClrDatagridColumnType = ClrDatagridColumnType.STRING;
+  public get colType() {
+    if (this._colType === ClrDatagridColumnType.NUMBER) {
+      return 'number';
```
Comment (youdz) [inline src/clr-angular/data/datagrid/datagrid-column.ts]: No offense, but this looks a bit silly to me. You receive a string as an input, convert it to an enum, and then you re-convert it to a string when you use it in your own template? Either just use strings everywhere, or embrace enums and don't convert back to strings.

To be clear, I think the best solution is simply to use a string union type rather than an enum:
```
@Input('clrDgColType') colType: "string" | "number" = "string";
```
No getter or setter needed, and the default works fine. If the consumer gives us anything other than these two strings, AoT compilation would fail with a clear message.

File: src/clr-angular/data/datagrid/datagrid-column.ts [hunk 4]
```diff
@@ -341,25 +378,38 @@ export class ClrDatagridColumn<T = any> extends DatagridFilterRegistrar<T, Datag
...
   public get filterValue() {
-    return this.filter.value;
+    if (this.filter instanceof DatagridStringFilterImpl || this.filter instanceof DatagridNumericFilterImpl) {
+      return this.filter.value;
+    }
   }
 
   @Input('clrFilterValue')
-  public set updateFilterValue(newValue: string) {
+  public set updateFilterValue(newValue: string | [number, number]) {
```
Comment (youdz) [inline src/clr-angular/data/datagrid/datagrid-column.ts]: Note to self for future updates: we'll need to find a way to type the column so that we can infer the type of this setter based on the `colType` of the column.

Don't worry about it for this PR.

File: src/clr-angular/data/datagrid/datagrid-column.ts [hunk 5]
```diff
@@ -44,10 +48,15 @@ let nbCount: number = 0;
             <ng-content select="clr-dg-filter, clr-dg-string-filter"></ng-content>
```
Comment (youdz) [inline src/clr-angular/data/datagrid/datagrid-column.ts]: Just realized, you need to update this selector too to add `clr-dg-numeric-filter`. Otherwise you won't project custom filters in the right place.

**Comment by youdz** [comment]
To answer your question of "Why is the filter icon on the wrong side?!?", I looked quickly at our CSS and realized we have some CSS from an external contribution that uses flex order to put the filter last:
```css
clr-dg-filter,
clr-dg-string-filter {
  display: flex;
  order: 99;
  margin-left: auto;
}
```

This doesn't respect our guidelines, styling Angular elements directly, so I'm okay with you stripping it and reordering the DOM so that the DOM order is the same as the display order. Or whatever solution you think is cleanest.

**Comment by amellnik** [comment]
Would it be possible to create a new feature branch for this?  I'm planning on working through the suggestions above this week and it seems like it would be nice to build it on top of https://github.com/vmware/clarity/commit/6310d4b79adad171c167e863d32870b43fc0cbb3.

**Comment by youdz** [comment]
Since you haven't committed anything to it yet, I'll just update the existing topic branch so it's based on latest master. Please make sure you rebase your changes on top of today's master before updating the PR. 👍

**Comment by amellnik** [comment]
I'm working on the documentation and the most natural way to demo the numeric filters often runs into https://github.com/vmware/clarity/issues/2683 since they are wide and you often want to add them to the first column.  Should I go ahead and add them normally or try to work around the issue by reordering columns to put the filters in the middle of the grid?

**Comment by hippee-lee** [comment]
Hey @amellnik - it looks like your changes affect the public Clarity API. You should be able to fix it by running `npm run golden:fix`.

**Comment by amellnik** [comment]
I just noticed that while the internationalization appears to work locally for testing, I might have messed something up since it shows up like this when I build, publish, then use from a different project:  

screen shot 2019-02-27 at 1 47 05 pm

**Comment by amellnik** [comment]
I just discovered one other issue that probably needs to be addressed -- in `clrDgRefresh` the `filter` array looks like this:

screen shot 2019-02-27 at 4 16 50 pm

for a built-in string filter, a built-in numeric filter, and a custom string filter respectively.  We should probably include the value of `clrDgColType` to make it clear what type the built-in filters are.

**Comment by gnomeontherun** [comment]
@amellnik Just checking in, how are things?

**Comment by amellnik** [comment]
@gnomeontherun I will fix the two minor issues I mentioned above, but I'm mostly just waiting on the design.  Is there anything else I should be working on?

We're using this for an internal tool now and haven't ran into any other issues with it.

**Comment by mathisscott** [comment]
@colinreedmiller and @lil-kim 
Could one of you take a look at this to get it unblocked? Thanks.

**Comment by sdurnov** [comment]
@amellnik, given that Clarity filters aren't closing automatically and you can open filters for several columns simultaneously, wouldn't it be better to place fields vertically, so they won't overlap with other filters?

**Comment by amellnik** [comment]
@sdurnov - I don't have any strong feelings about the layout, and would be open to stacking the inputs vertically.  I think the function of the filter is a bit more clear when they are horizontal since many users may be used to seeing ranges written as `[lower_limit, upper_limit]` or similar, but having the wide filters is annoying especially given https://github.com/vmware/clarity/issues/1262#issuecomment-420265815.

**Comment by amellnik** [comment]
image

This now matches @colinreedmiller's design except that it still has a X to close the filter and does show the up/down arrows.  I saw a few examples for removing the arrows in various browsers [here](https://stackoverflow.com/questions/3790935/can-i-hide-the-html5-number-input-s-spin-box) but they seem a bit hacky.  Do we want to do this, and if so, is there a recommended method?

**Comment by colinreedmiller** [comment]
We do not want to attempt to manage the increment / de-increment arrows. They will be provided or not depending on the browser.

Design spec here:

**Comment by colinreedmiller** [comment]
also: This design is for the numeric range implementation only. We are going to hold off on the operator / value version for now in deference to the advanced datagrid filtering currently in the cue: https://github.com/vmware/clarity/issues/1632

**Comment by amellnik** [comment]
I think this now matches the design, let me know if I missed anything or there's anything else I should do!

**Comment by amellnik** [comment]
@gnomeontherun @colinreedmiller Just checking in -- is there anything else I should do on this?

**Comment by gnomeontherun** [comment]
I've started a review, but have not completed it yet. I'm tied up at ng-conf right now but am trying to get it done asap!

File: src/website/src/app/documentation/demos/datagrid/built-in-filters/built-in-filters.html [hunk 6]
```diff
@@ -37,6 +36,18 @@ <h4>String filter</h4>
     entering the exact number of the pokemon he's interested in.
 </p>
 
+<h4>Numeric filter</h4>
+
+<p>
+    The "Wins" column demonstrates a built-in numeric filter.  You provide the function logic and
```
Comment (gnomeontherun) [inline src/website/src/app/documentation/demos/datagrid/built-in-filters/built-in-filters.html]: Update to

> Another built-in filter is the numeric filter, which allows you to filter a column by a minimum and/or maximum numeric value. The "Wins" column demonstrates the numeric filter.  You provide the function logic and the user can optionally enter high and low limits for elements in the column.  In this case, use a <code class="clr-code">&lt;clr-dg-numeric-filter&gt;</code> component and pass the filter to the <code class="clr-code">[clrDgNumericFilter]</code> property.

File: src/website/src/app/documentation/demos/datagrid/built-in-filters/built-in-filters.html [hunk 7]
```diff
@@ -37,6 +36,18 @@ <h4>String filter</h4>
...
+    the user can optionally enter high and low limits for elements in the column.  In this case,
+    use a <code class="clr-code">&lt;clr-dg-numeric-filter&gt;</code> component and pass the filter
+    to the <code class="clr-code">[clrDgNumericFilter]</code> property.
+</p>
+<clr-code-snippet [clrCode]="examples.numericFilterInterface" clrLanguage="typescript"></clr-code-snippet>
+<!-- TODO: Is this next example really needed? -->
+<!-- <clr-code-snippet [clrCode]="examples.numericFilterInput" clrLanguage="html"></clr-code-snippet> -->
```
Comment (gnomeontherun) [inline src/website/src/app/documentation/demos/datagrid/built-in-filters/built-in-filters.html]: Yes I think we should include this line, please display it.

File: src/dev/src/app/datagrid/string-filtering/string-filtering.html [hunk 8]
```diff
@@ -30,13 +30,24 @@ <h4>String filter</h4>
...
+<p>
+    The ID column demonstrates a built-in numeric filter.  You provide the function logic and the user can
+    optionally enter high and low limits for elements in the column.  In this case, pass a <code class="clr-code">&lt;clr-dg-numeric-filter&gt;</code> component.
 </p>
 
 <clr-datagrid>
-  <clr-dg-column>User ID</clr-dg-column>
   <clr-dg-column>Name
   </clr-dg-column>
+  <!-- <clr-dg-column [clrDgColType]="'number'" [clrDgField]="'id'">User ID</clr-dg-column> -->
```
Comment (gnomeontherun) [inline src/dev/src/app/datagrid/string-filtering/string-filtering.html]: Clean up.

File: src/dev/src/app/datagrid/string-filtering/string-filtering.html [hunk 9]
```diff
@@ -18,9 +18,9 @@ <h2>Built-in filters</h2>
 <h4>String filter</h4>
```
Comment (gnomeontherun) [inline src/dev/src/app/datagrid/string-filtering/string-filtering.html]: Would you mind doing the same file renames like you did on the website from `string-filtering` to `built-in-filters`?

File: src/dev/src/app/datagrid/binding-properties/binding-properties.html [hunk 10]
```diff
@@ -16,20 +16,27 @@ <h2>Binding model properties to columns</h2>
...
+  <clr-dg-column clrDgField="id" clrDgColType="number">User ID</clr-dg-column>
+  <clr-dg-column clrDgField="name" [(clrFilterValue)]="nameFilter">Name</clr-dg-column>
+  <clr-dg-column clrDgField="pokemon.name">Pokemon</clr-dg-column>
+  <clr-dg-column>Favorite color</clr-dg-column>
 
   <clr-dg-row *clrDgItems="let user of users">
+    <clr-dg-cell>{{user.creation | date}}</clr-dg-cell>
     <clr-dg-cell>{{user.id}}</clr-dg-cell>
     <clr-dg-cell>{{user.name}}</clr-dg-cell>
-    <clr-dg-cell>{{user.creation | date}}</clr-dg-cell>
```
Comment (gnomeontherun) [inline src/dev/src/app/datagrid/binding-properties/binding-properties.html]: Can you put the `creation` column back in the same place?

File: src/clr-angular/utils/i18n/common-strings.service.ts [hunk 11]
```diff
@@ -29,6 +29,8 @@ export class ClrCommonStringsService implements ClrCommonStrings {
   danger = 'Error';
   rowActions = 'Available actions';
   pickColumns = 'Show or hide columns';
+  min = 'Min';
```
Comment (gnomeontherun) [inline src/clr-angular/utils/i18n/common-strings.service.ts]: I think we need to make these more clear `Minimum value` and `Maximum value`.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts [hunk 12]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Observable } from 'rxjs';
+import { Subject } from 'rxjs';
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts]: Merge these two imports.

Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts]: Merge these two imports

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    // We have a filter function in case someone wants to implement a numeric
+    // filter that always passes nulls or similar
+    return this.filterFn.accepts(item, this._low, this._high);
+  }
+
+  public get state() {
+    if (this.filterFn instanceof DatagridPropertyNumericFilter) {
+      return {
+        property: this.filterFn.prop,
+        // TODO: Should this return value: [this._low, this._high] instead?
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts]: I think this is good as it is, that gives an explicit property for the consumer to have instead of having to unpack an array.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.spec.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    });
+
+    it('filters numbers that are below the low limit', function() {
+      expect(fullFilter.accepts(0)).toBe(true);
+      fullFilter.low = 1;
+      expect(fullFilter.accepts(0)).toBe(false);
+      expect(fullFilter.accepts(2)).toBe(true);
+      fullFilter.low = null;
+      expect(fullFilter.accepts(0)).toBe(true);
+    });
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.spec.ts]: Can we add another test that has both high and low limits set and check the bounds?

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                it prevents NgControlStatus and other directives automatically added by Angular
+                on inputs with NgModel from freaking out because of their host binding changing
+                mid-change detection when the input is destroyed.
+            -->
+            <input #input_low type="number" name="low" [(ngModel)]="low" *ngIf="open"
+                (keyup.enter)="close()" (keyup.escape)="close()" style="width: 78px"
+                [placeholder]="commonStrings.min"/>
+                <span class="datagrid-filter-input-spacer"></span>
+            <input #input_high type="number" name="high" [(ngModel)]="high" *ngIf="open"
+                (keyup.enter)="close()" (keyup.escape)="close()" style="width: 78px"
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: Can we rework the styling so there isn't an explicit width set on the input? That would move some CSS into the datagrid file. I would rather see us have something like a flex element that can scale in case popover sizes are ever changed and it surprises me that we don't have the `clr-input` class on here either for the styling.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+  providers: [{ provide: CustomFilter, useExisting: DatagridNumericFilter }],
+  template: `
+        <clr-dg-filter [clrDgFilter]="registered" [(clrDgFilterOpen)]="open">
+            <!--
+                Even though this *ngIf looks useless because the filter container already has one,
+                it prevents NgControlStatus and other directives automatically added by Angular
+                on inputs with NgModel from freaking out because of their host binding changing
+                mid-change detection when the input is destroyed.
+            -->
+            <input #input_low type="number" name="low" [(ngModel)]="low" *ngIf="open"
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: I believe we'll need to add `[aria-label]="commonStrings.min"` to this and the next input (max for that one of course) for a11y. There is no label otherwise and VoiceOver just says 'Stepper' but without context to what it does.

File: src/clr-angular/utils/i18n/common-strings.interface.ts [hunk 17]
```diff
@@ -85,4 +85,12 @@ export abstract class ClrCommonStrings {
    * Datagrid: pick columns
    */
   pickColumns?: string;
+  /**
+   * Datagrid numeric filter: min
+   */
+  min?: string;
```
Comment (gnomeontherun) [inline src/clr-angular/utils/i18n/common-strings.interface.ts]: When you rebase onto `master`, we'll need to add these into the internationalization documentation as well.

File: src/clr-angular/data/datagrid/datagrid-column.ts [hunk 18]
```diff
@@ -122,6 +130,14 @@ export class ClrDatagridColumn<T = any> extends DatagridFilterRegistrar<T, Datag
     this._sortSubscription.unsubscribe();
   }
 
+  /*
+     * What type is this column?  This defaults to STRING, but can also be
+     * set to NUMBER.  Unsupported types default to STRING. Users can set it
+     * via the [clrDgColType] input by setting it to 'string' or 'number'.
+     */
+
+  @Input('clrDgColType') colType: 'string' | 'number' = 'string';
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/datagrid-column.ts]: This is fine for now, thinking we'd want to setup an enum to track these types better in the future. Maybe make a comment todo, but fine for now.

File: src/clr-angular/data/datagrid/datagrid-column.ts [hunk 19]
```diff
@@ -319,25 +339,38 @@ export class ClrDatagridColumn<T = any> extends DatagridFilterRegistrar<T, Datag
...
+        this.filter.value = newValue;
+      }
     }
   }
 
-  public set filterValue(newValue: string) {
-    this.updateFilterValue = newValue;
-    this.filterValueChange.emit(this.filter.value);
+  public set filterValue(newValue: string | [number, number]) {
+    if (this.filter instanceof DatagridStringFilterImpl || this.filter instanceof DatagridNumericFilterImpl) {
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/datagrid-column.ts]: Why was this changed?

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-property-numeric-filter.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
```
Comment (gnomeontherun) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-property-numeric-filter.ts]: Noticed we'll need all the copyrights updated for these files to 2019. Current master has a script to do this for you though, so you could wait until rebasing.

File: src/dev/src/app/datagrid/binding-properties/binding-properties.html (hunk 10 above)
Comment (amellnik) [reply src/dev/src/app/datagrid/binding-properties/binding-properties.html]: I rearranged it because otherwise we run into this bug (https://github.com/vmware/clarity/issues/1262#issuecomment-420265815 and https://github.com/vmware/clarity/issues/2683)

image

It looks like this may take a while to get resolved -- should I add one more additional column in the first position or what?

File: src/clr-angular/utils/i18n/common-strings.service.ts (hunk 11 above)
Comment (amellnik) [reply src/clr-angular/utils/i18n/common-strings.service.ts]: At the current input widths these placeholders would get cut off.  Is that OK?

Comment (amellnik) [reply src/clr-angular/utils/i18n/common-strings.service.ts]: `Min value` and `Max value` mostly fit if that works:

image

File: src/clr-angular/data/datagrid/datagrid-column.ts (hunk 19 above)
Comment (amellnik) [reply src/clr-angular/data/datagrid/datagrid-column.ts]: `value` is not part of `ClrDatagridFilterInterface` but is part of the individual string and numeric filters.  Originally I was going to have the state of the numeric filter be called `values` but I switched it to `value` to match.  Should we make this part of the interface?  I assume any future built-in filter will have some sort of internal state that we can refer to with `value`.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts (hunk 15 above)
Comment (amellnik) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: I moved the fixed width to the CSS.  I don't know much about design -- should we have breakpoints for this now or is that a TODO?

File: src/clr-angular/utils/i18n/common-strings.interface.ts (hunk 17 above)
Comment (amellnik) [reply src/clr-angular/utils/i18n/common-strings.interface.ts]: Keeping open for the internationalization todo.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-property-numeric-filter.ts (hunk 20 above)
Comment (amellnik) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-property-numeric-filter.ts]: Cool, holding off for now.

File: src/dev/src/app/datagrid/binding-properties/binding-properties.html (hunk 10 above)
Comment (gnomeontherun) [reply src/dev/src/app/datagrid/binding-properties/binding-properties.html]: Oh yes, that is something that is being actively addressed. I see why you did it now, and you can leave it.

File: src/clr-angular/utils/i18n/common-strings.service.ts (hunk 11 above)
Comment (gnomeontherun) [reply src/clr-angular/utils/i18n/common-strings.service.ts]: I can agree with that. It might also depend on the browser.

File: src/clr-angular/data/datagrid/datagrid-column.ts (hunk 19 above)
Comment (gnomeontherun) [reply src/clr-angular/data/datagrid/datagrid-column.ts]: Ok, makes sense. I think we can leave it as is. We have some tech debt on making filters more generically supportable but that is a separate task!

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts (hunk 15 above)
Comment (gnomeontherun) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: No breakpoint at the moment, just would like to get the style attribute out from here.

**Comment by amellnik** [comment]
I think this should be ready for rebasing but let me know if there's anything else I should do.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.spec.ts [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    it('only allows values within the range when both filters are set', function() {
+      fullFilter.low = 1;
+      fullFilter.high = 10;
+      expect(fullFilter.accepts(0)).toBe(false);
+      expect(fullFilter.accepts(5)).toBe(true);
+      expect(fullFilter.accepts(12)).toBe(false);
+    });
+
+    // TODO: why does this throw an error on build?
+    // it('exposes state', function() {
```
Comment (coryrylan) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.spec.ts]: What was the error in the build?

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
```
Comment (coryrylan) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.ts]: nit: remove outdated comment

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.spec.ts [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+      context.clarityDirective.value = [null, 10];
+      expect(filtersInstance.getActiveFilters().length).toBe(1);
+      expect(filtersInstance.getActiveFilters()[0]).toBe(context.clarityDirective.filter);
+    });
+
+    it('registers itself as a CustomFilter provider', function() {
+      expect(context.testComponent.customFilter).toBe(context.clarityDirective);
+    });
+
+    // TODO: Should this count the inputs?
```
Comment (coryrylan) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.spec.ts]: I don't think that is necessary, we can use gemini to make a screenshot of the filter to make sure everything is looks correct.

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+   * We need the actual input element to automatically focus on it
+   */
+  @ViewChild('input_low') public input: ElementRef;
+
+  /**
+   * We grab the ClrDatagridFilter we wrap to register this StringFilter to it.
+   */
+  @ViewChild(ClrDatagridFilter) public filterContainer: ClrDatagridFilter<T>;
+  ngAfterViewInit() {
+    this.filterContainer.openChanged.subscribe((open: boolean) => {
```
Comment (coryrylan) [inline src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: Should unsubscribe in ngOnDestroy

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.spec.ts (hunk 21 above)
Comment (amellnik) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter-impl.spec.ts]: There was actually nothing wrong with the test, I missed something in the `state` function.  The test now works.

File: package-lock.json [hunk 25]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@clr/clarity",
-  "version": "1.0.4",
+  "version": "1.0.5",
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
```
[the remark this answers is not in the document]
Comment (amellnik) [reply package-lock.json]: I think I actually corrected these now!  (They were left over from a hacky publish to an internal npm proxy so that we could start using the feature).

File: src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts (hunk 24 above)
Comment (amellnik) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: Like this?  https://github.com/vmware/clarity/pull/2996/files#diff-69d97a05a7a4f27670ea86537ac65355R83

Comment (Jinnie) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: We're usually using the unsubscribe on the Subscription object
```
this.subscription = this.filterContainer.openChanged.subscribe(...
...
ngOnDestroy() {
  this.subscription.unsubscribe();
}
```
and lately we use array of subscriptions even for single subscriptions, like this:
```
private subscriptions: Subscription[] = [];
....
this.subscriptions.push(this.filterContainer.openChanged.subscribe(...
...
ngOnDestroy() {
  this.subscriptions.forEach(sub => {
    sub.unsubscribe();
  });
}
```
so that it's easy to manage multiple subscriptions if some are added in future.

Comment (amellnik) [reply src/clr-angular/data/datagrid/built-in/filters/datagrid-numeric-filter.ts]: I'll switch it to the array version this evening and also re-run `golden:fix` (why the tests failed).  Thanks!

**Comment by coryrylan** [comment]
This looks good, I think it could be rebased and then once rebased update the copyright dates on the altered files to 2019, as well as documenting the new entries for internationalization.

**Comment by gnomeontherun** [comment]
From a pragmatic standpoint, it might be easiest to just rebase and make a new PR to master and close this out. That way we don't have to do too much git fu to move forward.

**Comment by amellnik** [comment]
Closing in favor of #3397
